src/V3CfgLiveVariables.cpp

```diff
@@ -0,0 +1,233 @@
+//
+// Classical data flow analysis computing live variables input to a CFG
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Cfg.h"
+
+#include <limits>
+#include <memory>
+#include <unordered_map>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+template <bool T_Scoped>
+class CfgLiveVariables final : VNVisitorConst {
+    // TYPES
+    using Variable = std::conditional_t<T_Scoped, AstVarScope, AstVar>;
+
+    struct BlockState final {
+        std::unordered_set<Variable*> m_gen;
+        std::unordered_set<Variable*> m_kill;
+        std::unordered_set<Variable*> m_liveIn;
+        std::unordered_set<Variable*> m_liveOut;
+        bool m_isOnWorkList = false;
+        bool m_wasProcessed = false;  // Already processed at least once
+    };
+
+    // STATE
+    const ControlFlowGraph& m_cfg;  // The CFG beign analysed
+    // State for each block
+    BasicBlockMap<BlockState> m_blockState = m_cfg.makeBasicBlockMap<BlockState>();
+    BlockState* m_currp = nullptr;  // State of current block being analysed
+    bool m_abort = false;  // Abort analysis - unhandled construct
+
+    // METHODS
+    static Variable* getTarget(const AstNodeVarRef* refp) {
+        // TODO: remove the useless reinterpret_casts when C++17 'if constexpr' actually works
+        if VL_CONSTEXPR_CXX17 (T_Scoped) {
+            return reinterpret_cast<Variable*>(refp->varScopep());
+        } else {
+            return reinterpret_cast<Variable*>(refp->varp());
+        }
+    }
+
+    // This is to match DFG, but can be extended independently - eventually should handle all
+    static bool isSupportedPackedDType(const AstNodeDType* dtypep) {
+        dtypep = dtypep->skipRefp();
+        if (const AstBasicDType* const typep = VN_CAST(dtypep, BasicDType)) {
+            return typep->keyword().isIntNumeric();
+        }
+        if (const AstPackArrayDType* const typep = VN_CAST(dtypep, PackArrayDType)) {
+            return isSupportedPackedDType(typep->subDTypep());
+        }
+        if (const AstNodeUOrStructDType* const typep = VN_CAST(dtypep, NodeUOrStructDType)) {
+            return typep->packed();
+        }
+        return false;
+    }
+
+    // Check and return if variable is incompatible
+    bool incompatible(Variable* varp) {
+        if (!isSupportedPackedDType(varp->dtypep())) return true;
+        AstVar* astVarp = nullptr;
+        // TODO: remove the useless reinterpret_casts when C++17 'if constexpr' actually works
+        if VL_CONSTEXPR_CXX17 (T_Scoped) {
+            astVarp = reinterpret_cast<AstVarScope*>(varp)->varp();
+        } else {
+            astVarp = reinterpret_cast<AstVar*>(varp);
+        }
+        if (astVarp->ignoreSchedWrite()) return true;
+        return false;
+    }
+
+    void updateGen(AstNode* nodep) {
+        m_abort |= nodep->exists([&](AstNodeVarRef* refp) {
```

Comment:
the `|=` won't short circuit so we do useless work. Similar below.

```suggestion
    void updateGen(AstNode* nodep) {
        if (m_abort) return;  // Short circuit
        m_abort |= nodep->exists([&](AstNodeVarRef* refp) {
```